synapse/rest/media/v1/media_repository.py

```diff
@@ -476,7 +498,8 @@ async def _download_remote_file(
                 time_now_ms=self.clock.time_msec(),
                 upload_name=upload_name,
                 media_length=length,
-                filesystem_id=file_id,
+                filesystem_id=file_info.file_id,
+                update=update_store,
             )
```

Comment:
This is what the PR fixes fundamentally, it will reuse the file ID and download to it again when it is missing from all stores.

Reply:
My questions are:
* What happens if a user tries to /download remote media while we are *re*downloading it? Will they get a truncated copy?
* What happens if two media repo workers try to *re*download the same remote media? What will be the final contents of the media on disk? What will `StorageProvider`s observe?

Reply:
> What happens if a user tries to /download remote media while we are redownloading it? Will they get a truncated copy?

Not entirely sure, it depends how the file is flushed to disk.

There is `self.remote_media_linearizer` which makes sure that two calls for the same download on the same process will not happen simultaneously, but this is not the case when two media repo workers try to write and read on disk at the same time, in that instance it would read from a truncated file, methinks.

This would then only happen once, as there's only a small timewindow where the initial contents of a remote re(!)download are flushed before the rest is flushed, and only if a download is called at another media repo (which itself is not engaging in redownloading at that time) before the full contents is flushed to disk.

In other words, yes, it can race if multiple media repo workers exist to write to the same directory, if media is redownloaded (only happens when the media repo folder is manually deleted), and the other media worker only "just now" starts looking for the file, is this acceptable?

I don't know how to fix this issue, other than maybe issuing a database `DELETE` if it thinks the media is non-existent, before trying to download it "the normal way" (with unique violation exceptions preventing racing)

> What happens if two media repo workers try to redownload the same remote media? What will be the final contents of the media on disk? What will StorageProviders observe?

The final contents (on disk) will be the write-truncate flush of the last media repo worker that touched it, storage providers would see two store requests (and some fetch requests to check if the media is there)

Reply:
Thinking about this, I'm realizing that deleting-record-then-downloading is probably a better way to solve this just overall, I'll rework this to that approach. :+1: